pkg/fleet/installer/repository/link_win.go

```diff
@@ -0,0 +1,217 @@
+	if err != nil && !errors.Is(err, syscall.ERROR_ALREADY_EXISTS) {
+		return err
+	}
+
+	linkp, err := syscall.UTF16FromString(link)
+	if err != nil {
+		return err
+	}
+
+	fd, err := syscall.CreateFile(&linkp[0], syscall.GENERIC_WRITE, 0, nil, syscall.OPEN_EXISTING,
+		syscall.FILE_FLAG_OPEN_REPARSE_POINT|syscall.FILE_FLAG_BACKUP_SEMANTICS, 0)
+	if err != nil {
+		return err
+	}
+	defer syscall.CloseHandle(fd)
+
+	buflen := uint32(rdb.header.ReparseDataLength) + uint32(unsafe.Sizeof(rdb.header))
+	var bytesReturned uint32
+	return syscall.DeviceIoControl(fd, windows.FSCTL_SET_REPARSE_POINT,
+		(*byte)(unsafe.Pointer(&rdb.header)), buflen, nil, 0, &bytesReturned, nil)
+}
+
+func createSymbolicLink(link string, target *reparseData, isrelative bool) error {
+	var buf *symbolicLinkReparseBuffer
+	buflen := uint16(unsafe.Offsetof(buf.PathBuffer)) + target.pathBuffeLen() // see ReparseDataLength documentation
+	byteblob := make([]byte, buflen)
+	buf = (*symbolicLinkReparseBuffer)(unsafe.Pointer(&byteblob[0]))
+	buf.SubstituteNameOffset = target.substituteName.offset
+	buf.SubstituteNameLength = target.substituteName.length
+	buf.PrintNameOffset = target.printName.offset
+	buf.PrintNameLength = target.printName.length
+	if isrelative {
+		buf.Flags = SYMLINK_FLAG_RELATIVE
+	}
+	pbuflen := len(target.pathBuf)
+	copy((*[2048]uint16)(unsafe.Pointer(&buf.PathBuffer[0]))[:pbuflen:pbuflen], target.pathBuf)
+
+	var rdb _REPARSE_DATA_BUFFER
+	rdb.header.ReparseTag = syscall.IO_REPARSE_TAG_SYMLINK
+	rdb.header.ReparseDataLength = buflen
+	copy(rdb.detail[:], byteblob)
+
+	return createDirLink(link, &rdb)
+}
+
+func setReparsePoint(target, link string) error {
+	var t reparseData
+	t.addPrintName(target)
+	t.addSubstituteName(`\??\` + target)
+	return createSymbolicLink(link, &t, false)
+}
+
+func symlinkWithImpersonation(target, link string, fn func(target, link string) error) error {
+	err := windows.ImpersonateSelf(windows.SecurityImpersonation)
+	if err != nil {
+		return err
+	}
+	defer windows.RevertToSelf()
+
+	err = enableCurrentThreadPrivilege("SeCreateSymbolicLinkPrivilege")
+	if err != nil {
+		return fmt.Errorf(`could not enable "SeCreateSymbolicLinkPrivilege": %v`, err)
+	}
+
+	return fn(target, link)
+}
+
+func enableCurrentThreadPrivilege(privilegeName string) error {
+	ct := windows.CurrentThread()
+	var t windows.Token
+	err := windows.OpenThreadToken(ct, syscall.TOKEN_QUERY|windows.TOKEN_ADJUST_PRIVILEGES, false, &t)
+	if err != nil {
+		return err
+	}
+	defer syscall.CloseHandle(syscall.Handle(t))
+
+	var tp windows.Tokenprivileges
+
+	privStr, err := syscall.UTF16PtrFromString(privilegeName)
+	if err != nil {
+		return err
+	}
+	err = windows.LookupPrivilegeValue(nil, privStr, &tp.Privileges[0].Luid)
+	if err != nil {
+		return err
+	}
+	tp.PrivilegeCount = 1
+	tp.Privileges[0].Attributes = windows.SE_PRIVILEGE_ENABLED
+	return windows.AdjustTokenPrivileges(t, false, &tp, 0, nil, nil)
+}
+
+func atomicSymlink(oldname, newname string) error {
```

Comment:
some notes...

golang [`os.Symlink`](https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/os/file_windows.go;l=324) calls Windows [`CreateSymbolicLink`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinka)

`CreateSymbolicLink` calls
  - `NtCreateFile(..., FILE_SHARE_NONE, FILE_CREATE, ...)`
  - `NtFsControlFile(..., FSCTL_SET_REPARSE_POINT, ...)`
So the FILE_CREATE flag is why it fails if the link already exists, and why we can't use it here.

The `createDirLink()` implementation is very similar to `CreateSymbolicLink`

Neither are truly "atomic" though.

`FILE_SHARE_NONE` means nobody can access the link until we're done updating it, so there's a (very) small window for error.

to support existing links, `createDirLink()` calls `os.Mkdir(link)` and later procides `OPEN_EXISTING` instead of `FILE_CREATE`.
so there's a (very) small window for error, on initial creation, if the directory/link is accessed before we make it a link.
  - we could remove `os.Mkdir` and use `OPEN_ALWAYS` instead, which might trigger an error from `FILE_SHARE_NONE` on the other user instead of the other way around, I'm not sure if there are other considerations.

In any case, this code does allow us to ensure that if our process dies that we are left with a working link, which from previous discussion seems more important for our use case.

I tested that the link can be successfully updated when a file in the directory is currently open, and that it can be read after the link is updated.

Can you add some of these possible error cases as a comment on "atomicSymlink" so we can track that it differs from the Linux implementation, and highlight its main goal is to handle "if our process dies that we are left with a working link"?

Reply:
> it creates the symlink as a regular file

Ah, too bad, thanks for checking.